lib/sources.nix

```diff
@@ -26,14 +26,32 @@ rec {
     (type == "symlink" && lib.hasPrefix "result" baseName)
   );
 
-  cleanSource = builtins.filterSource cleanSourceFilter;
+  cleanSource = src: cleanSourceWith { filter = cleanSourceFilter; inherit src; };
+
+  # Like `builtins.filterSource`, except it will compose with itself,
+  # allowing you to chain multiple calls together without any
+  # intermediate copies being put in the nix store.
+  cleanSourceWith = { filter, src }:
+    let
+      isFiltered = src ? _isLibCleanSourceWith;
+      origSrc = if isFiltered then src.origSrc else src;
+      filter' = if isFiltered then name: type: filter name type && src.filter name type else filter;
+    in {
+      inherit origSrc;
+      filter = filter';
+      outPath = builtins.filterSource filter' origSrc;
+      _isLibCleanSourceWith = true;
+    };
 
   # Filter sources by a list of regular expressions.
   #
   # E.g. `src = sourceByRegex ./my-subproject [".*\.py$" "^database.sql$"]`
```

Comment:
IIRC this is actually a misleading example. I believe this will not recurse into directories unless they match these regexes. So in many cases this example would actually filter out everything and leave you with an empty folder. I need to double-check that this is the case but that's how it has appeared in my experience.

Reply:
I realize this is tangential to your actual change. Feel free to ignore while I look into opening a separate PR.

Reply:
I've confirmed this. I'll open a separate PR.